engine/immediate_task_conrrent_processor.go

```diff
@@ -219,6 +223,53 @@ func (w *immediateTaskConcurrentProcessor) processWaitUntilTask(
 	return nil
 }
 
+func (w *immediateTaskConcurrentProcessor) applyStateFailureRecoveryPolicy(ctx context.Context,
+	task persistence.ImmediateTask,
+	prep persistence.PrepareStateExecutionResponse,
+	stateApiType xdbapi.StateApiType,
+	apiClient *xdbapi.APIClient) error {
+	stateRecoveryPolicy := xdbapi.AsyncStateConfigStateFailureRecoveryInfo{
+		Policy: xdbapi.FAIL_PROCESS_ON_STATE_FAILURE,
+	}
+	if prep.Info.StateConfig != nil && prep.Info.StateConfig.StateFailureRecoveryInfo != nil {
+		stateRecoveryPolicy = *prep.Info.StateConfig.StateFailureRecoveryInfo
+	}
+	switch stateRecoveryPolicy.Policy {
+	case xdbapi.FAIL_PROCESS_ON_STATE_FAILURE:
+		resp, errStopProcess := w.store.StopProcess(ctx, persistence.StopProcessRequest{
+			Namespace:       prep.Info.Namespace,
+			ProcessId:       prep.Info.ProcessId,
+			ProcessStopType: xdbapi.FAIL,
+		})
+
+		if errStopProcess != nil {
+			return errStopProcess
+		}
+		if resp.NotExists {
+			// this should not happen
+			return fmt.Errorf("process does not exist")
+		}
+	case xdbapi.PROCEED_TO_CONFIGURED_STATE:
+		w.store.MoveProcessToState(ctx, persistence.MoveProcessToStateRequest{
+			Namespace:          prep.Info.Namespace,
+			ProcessExecutionId: task.ProcessExecutionId,
+			SourceStateExecutionId: persistence.StateExecutionId{
+				StateId:         task.StateId,
+				StateIdSequence: task.StateIdSequence,
+			},
+			SourceFailedStateApi:   stateApiType,
```

Comment:
I think we also need to store the lastFailure to the source state execution for debugging